src/main/java/com/google/jenkins/plugins/k8sengine/KubernetesEngineBuilder.java

```diff
@@ -180,26 +200,125 @@ public void perform(
     // generate a kubeconfig for the cluster
     KubeConfig kubeConfig = KubeConfig.fromCluster(projectId, cluster);
 
-    // run kubectl apply
-    KubectlWrapper.runKubectlCommand(
+    JenkinsRunContext context =
         new JenkinsRunContext.Builder()
             .workspace(workspace)
             .launcher(launcher)
             .taskListener(listener)
             .run(run)
-            .build(),
+            .build();
+    KubectlWrapper kubectl = new KubectlWrapper(context, kubeConfig);
+    kubectl.runKubectlCommand(
+        context,
         kubeConfig,
         "apply",
-        ImmutableList.<String>of("-f", workspace.child(manifestPattern).getRemote()));
+        ImmutableList.<String>of("-f", workspace.child(manifestPattern).getRemote(), "-o", "yaml"));
+
+    if (verifyDeployments) {
+      verified = verify(kubectl, manifestPattern, workspace, listener.getLogger());
+      if (!verified) {
+        throw new AbortException("Kubernetes objects could not be verified.");
+      }
+    }
 
-    // run the after build step if it exists
-    // NOTE(craigatgoogle): Due to the reflective way this class is created, initializers aren't
-    // run, so we still have to check for null.
     if (afterBuildStep != null) {
       afterBuildStep.perform(kubeConfig, run, workspace, launcher, listener);
     }
   }
 
+  private Set<String> getKindsToVerify() {
+    return ImmutableSet.of("Deployment");
+  }
+
+  /*
+   * verify the application of the supplied manifest objects and return true if it succeeded
+   */
+  private boolean verify(
+      KubectlWrapper kubectl, String manifestPattern, FilePath workspace, PrintStream consoleLogger)
+      throws AbortException, InterruptedException, IOException {
+
+    LOGGER.log(
+        Level.INFO,
+        String.format(
+            "GKE verifying deployment to, projectId: %s cluster: %s zone: %s manifests: %s",
+            projectId, clusterName, zone, workspace.child(manifestPattern)));
+
+    consoleLogger.println(
+        String.format("Verifying manifests: %s", workspace.child(manifestPattern)));
+
+    Manifests manifests = Manifests.fromFile(workspace.child(manifestPattern));
+
+    // Filter by the kinds of manifests being verified.
+    List<Manifests.ManifestObject> manifestObjects =
+        manifests.getObjectManifestsOfKinds(getKindsToVerify());
+
+    consoleLogger.println(String.format("Verifying %d objects.", manifestObjects.size()));
+
+    final List<VerificationTask> tasks =
+        manifestObjects
+            .stream()
+            .map((mo) -> new VerificationTask(kubectl, mo, consoleLogger))
+            .collect(Collectors.toList());
+    ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(0);
```

Comment:
This seems like a lot of manual heavy lifting used for palatalization. Perhaps the amount of code used here could be reduced by using an existing library like ForkJoinPool: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html

Reply:
I'll take a look.  The scheduler allows for scheduling periodic tasks (polling) and timing out.  It seemed like a good fit.  The rest of this code is map reduce to  map a manifest to a verification task and report and reduce the results.